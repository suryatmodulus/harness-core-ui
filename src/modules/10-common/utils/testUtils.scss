/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
