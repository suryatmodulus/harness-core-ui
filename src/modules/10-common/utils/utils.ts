/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}
