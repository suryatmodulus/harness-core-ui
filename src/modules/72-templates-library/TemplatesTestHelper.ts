/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { ResponseString, ResponsePageTemplateSummaryResponse } from 'services/template-ng'
import type { UseGetMockDataWithMutateAndRefetch } from '@common/utils/testUtils'

export const stageMockTemplatesInputYaml: ResponseString = {
  status: 'SUCCESS',
  data:
    'type: "Deployment"' +
    '\nspec:' +
    '\n  serviceConfig:' +
    '\n    serviceRef: "<+input>"' +
    '\n  infrastructure:' +
    '\n    infrastructureDefinition:' +
    '\n      type: "KubernetesDirect"' +
    '\n      spec:' +
    '\n        namespace: "<+input>"' +
    '\n  execution:' +
    '\n    steps:' +
    '\n    - step:' +
    '\n        identifier: "Step_1"' +
    '\n        type: "ShellScript"' +
    '\n        spec:' +
    '\n          source:' +
    '\n            type: "Inline"' +
    '\n            spec:' +
    '\n              script: "<+input>"' +
    '\n'
}

export const mockTemplatesInputYaml: ResponseString = {
  status: 'SUCCESS',
  data:
    'type: "HarnessApproval"\ntimeout: "<+input>"' +
    '\nspec:\n  approvalMessage: "<+input>"\n  approvers:\n    userGroups: "<+input>"\n    minimumCount: "<+input>"\n'
}

export const mockTemplates: ResponsePageTemplateSummaryResponse = {
  status: 'SUCCESS',
  data: {
    content: [
      {
        accountId: 'kmpySmUISimoRrJL6NL73w',
        orgIdentifier: 'default',
        projectIdentifier: 'Templateproject',
        identifier: 'manjutesttemplate',
        name: 'manju-test-template-qq-12344',
        description:
          'Flink is a versatile framework, supporting many different deployment scenarios in a mix and match fashion.',
        tags: { QAR: '', 'Internal 1': '', 'Canary 1': '', BLUE: '', 'Tag A': '' },
        yaml:
          'template:' +
          '\n    name: manju-test-template-qq-12344' +
          '\n    identifier: manjutesttemplate' +
          '\n    versionLabel: v4' +
          '\n    type: Step' +
          '\n    projectIdentifier: Templateproject' +
          '\n    orgIdentifier: default' +
          '\n    description: Flink is a versatile framework, supporting many different deployment scenarios in a mix and match fashion.' +
          '\n    tags:' +
          '\n        QAR: ""' +
          '\n        Internal 1: ""' +
          '\n        Canary 1: ""' +
          '\n        BLUE: ""' +
          '\n        Tag A: ""' +
          '\n    spec:' +
          '\n        type: HarnessApproval' +
          '\n        timeout: <+input>' +
          '\n        spec:' +
          '\n            approvalMessage: <+input>' +
          '\n            includePipelineExecutionHistory: true' +
          '\n            approvers:' +
          '\n                userGroups: <+input>' +
          '\n                minimumCount: <+input>' +
          '\n                disallowPipelineExecutor: false' +
          '\n            approverInputs:' +
          '\n                - name: "1"' +
          '\n                  defaultValue: "1"' +
          '\n                - name: ttt' +
          '\n                  defaultValue: ttt' +
          '\n        failureStrategies:' +
          '\n            - onFailure:' +
          '\n                  errors:' +
          '\n                      - Timeout' +
          '\n                  action:' +
          '\n                      type: MarkAsSuccess' +
          '\n',
        versionLabel: 'v4',
        templateEntityType: 'Step',
        childType: 'HarnessApproval',
        templateScope: 'project',
        version: 8,
        gitDetails: {},
        lastUpdatedAt: 1635626311830,
        stableTemplate: true
      }
    ],
    pageable: {
      sort: { sorted: true, unsorted: false, empty: false },
      pageSize: 20,
      pageNumber: 0,
      offset: 0,
      paged: true,
      unpaged: false
    },
    last: true,
    totalPages: 1,
    totalElements: 1,
    sort: { sorted: true, unsorted: false, empty: false },
    first: true,
    numberOfElements: 5,
    size: 20,
    number: 0,
    empty: false
  }
}

export const mockTemplatesSuccessResponse: UseGetMockDataWithMutateAndRefetch<ResponsePageTemplateSummaryResponse> = {
  loading: false,
  refetch: jest.fn(),
  mutate: jest.fn(),
  cancel: jest.fn(),
  data: mockTemplates
}
