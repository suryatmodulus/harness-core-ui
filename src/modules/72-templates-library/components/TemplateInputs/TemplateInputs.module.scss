.container {
  position: relative;
  min-height: 300px;
  .stageCard {
    margin: 0px 0px var(--spacing-xxlarge) 0px;
    padding: var(--spacing-5) 0px;
    background: var(--white);
  }
  .inputsContainer {
    :global {
      .bp3-form-group {
        > .bp3-label {
          color: var(--grey-600) !important;
        }
      }
    }
    .inputsCount {
      flex-shrink: 0;
    }
    .inputsCard {
      box-shadow: 0px 0px 1px rgba(40, 41, 61, 0.08), 0px 0.5px 2px rgba(96, 97, 112, 0.16);
      border-radius: 4px;
    }
  }
}
