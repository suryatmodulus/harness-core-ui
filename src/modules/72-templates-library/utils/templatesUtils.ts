import type { UseStringsReturn } from 'framework/strings'

export enum TemplateType {
  Step = 'Step',
  Stage = 'Stage',
  Pipeline = 'Pipeline',
  Service = 'Service',
  Infrastructure = 'Infrastructure',
  StepGroup = 'StepGroup',
  Execution = 'Execution'
}

export const AllTemplatesTypes = 'All'

export interface TemplateTypeOption {
  label: string
  value: string
  disabled?: boolean
}

export const getAllowedTemplateTypes = (getString: UseStringsReturn['getString']): TemplateTypeOption[] => [
  {
    label: getString('step'),
    value: TemplateType.Step
  },
  {
    label: getString('templatesLibrary.stageTemplate'),
    value: TemplateType.Stage
  },
  {
    label: getString('common.pipeline'),
    value: TemplateType.Pipeline,
    disabled: true
  },
  {
    label: getString('service'),
    value: TemplateType.Service,
    disabled: true
  },
  {
    label: getString('infrastructureText'),
    value: TemplateType.Infrastructure,
    disabled: true
  },
  {
    label: getString('stepGroup'),
    value: TemplateType.StepGroup,
    disabled: true
  },
  {
    label: getString('executionText'),
    value: TemplateType.Execution,
    disabled: true
  }
]
