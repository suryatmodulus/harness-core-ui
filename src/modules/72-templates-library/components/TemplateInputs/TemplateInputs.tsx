import React from 'react'
import {
  Color,
  Container,
  MultiTypeInputType,
  Text,
  Layout,
  Formik,
  getMultiTypeFromValue,
  PageError
} from '@wings-software/uicore'
import { parse } from 'yaml'
import { defaultTo, noop } from 'lodash-es'
import cx from 'classnames'
import { TemplateSummaryResponse, useGetTemplateInputSetYaml } from 'services/template-ng'
import type { StepType } from '@pipeline/components/PipelineSteps/PipelineStepInterface'
import { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { StepWidget } from '@pipeline/components/AbstractSteps/StepWidget'
import factory from '@pipeline/components/PipelineSteps/PipelineStepFactory'
import { PageSpinner, useToaster } from '@common/components'
import type { StageElementConfig, StepElementConfig, StageElementWrapperConfig } from 'services/cd-ng'
import MultiTypeDelegateSelector from '@common/components/MultiTypeDelegateSelector/MultiTypeDelegateSelector'
import { useStrings } from 'framework/strings'
import { StageForm } from '@pipeline/components/PipelineInputSetForm/PipelineInputSetForm'
import { TemplateType } from '@templates-library/utils/templatesUtils'
import NoResultsView from '@templates-library/pages/TemplatesPage/views/NoResultsView/NoResultsView'
import { getTemplateNameWithLabel } from '@pipeline/utils/templateUtils'
import css from './TemplateInputs.module.scss'
import stepCss from '@pipeline/components/PipelineSteps/Steps/Steps.module.scss'

export interface TemplateInputsProps {
  template: TemplateSummaryResponse
}

export const TemplateInputs: React.FC<TemplateInputsProps> = props => {
  const { template } = props
  const [inputSetTemplate, setInputSetTemplate] = React.useState<StepElementConfig | StageElementConfig | null>()
  const [count, setCount] = React.useState<number>(0)
  const { showError } = useToaster()
  const { getString } = useStrings()
  const allowableTypes = [MultiTypeInputType.FIXED, MultiTypeInputType.EXPRESSION, MultiTypeInputType.RUNTIME]

  const {
    data: templateInputYaml,
    error: inputSetError,
    refetch,
    loading
  } = useGetTemplateInputSetYaml({
    templateIdentifier: defaultTo(template.identifier, ''),
    queryParams: {
      accountIdentifier: defaultTo(template.accountId, ''),
      orgIdentifier: template.orgIdentifier,
      projectIdentifier: template.projectIdentifier,
      versionLabel: defaultTo(template.versionLabel, '')
    }
  })

  React.useEffect(() => {
    try {
      const templateInput = parse(templateInputYaml?.data || '')
      setCount((JSON.stringify(templateInput).match(/<\+input>/g) || []).length)
      setInputSetTemplate(templateInput)
    } catch (error) {
      showError(error.message, undefined, 'template.parse.inputSet.error')
    }
  }, [templateInputYaml?.data])

  return (
    <Container
      style={{ overflow: 'auto' }}
      padding={{ top: 'xlarge', left: 'xxlarge', right: 'xxlarge' }}
      className={css.container}
    >
      <Layout.Vertical>
        {loading && <PageSpinner />}
        {!loading && inputSetError && (
          <Container height={300}>
            <PageError
              message={defaultTo((inputSetError.data as Error)?.message, inputSetError.message)}
              onClick={() => refetch()}
            />
          </Container>
        )}
        {!loading && !inputSetError && !inputSetTemplate && (
          <Container height={300}>
            <NoResultsView minimal={true} text={getString('templatesLibrary.noInputsRequired')} />
          </Container>
        )}
        {!loading && !inputSetError && inputSetTemplate && (
          <Container className={css.inputsContainer}>
            <Layout.Vertical spacing={'xlarge'}>
              <Container>
                <Layout.Horizontal flex={{ alignItems: 'center' }} spacing={'xxxlarge'}>
                  <Text font={{ size: 'normal', weight: 'bold' }} color={Color.GREY_800}>
                    {getTemplateNameWithLabel(template)}
                  </Text>
                  <Text className={css.inputsCount} font={{ size: 'small' }}>
                    {getString('templatesLibrary.inputsCount', { count })}
                  </Text>
                </Layout.Horizontal>
              </Container>
              <Formik<StepElementConfig | StageElementWrapperConfig>
                onSubmit={noop}
                initialValues={
                  template.templateEntityType === TemplateType.Step
                    ? (inputSetTemplate as StepElementConfig)
                    : ({ stage: inputSetTemplate } as StageElementWrapperConfig)
                }
                formName="templateInputs"
                enableReinitialize={true}
              >
                {formikProps => {
                  return (
                    <>
                      {template.templateEntityType === TemplateType.Stage && (
                        <StageForm
                          template={formikProps.values as StageElementWrapperConfig}
                          allValues={formikProps.values as StageElementWrapperConfig}
                          path={'stage'}
                          readonly={true}
                          viewType={StepViewType.InputSet}
                          allowableTypes={allowableTypes}
                          hideTitle={true}
                          stageClassName={css.stageCard}
                        />
                      )}
                      {template.templateEntityType === TemplateType.Step && (
                        <Container
                          className={css.inputsCard}
                          background={Color.WHITE}
                          padding={'large'}
                          margin={{ bottom: 'xxlarge' }}
                        >
                          <StepWidget<Partial<StepElementConfig>>
                            factory={factory}
                            initialValues={formikProps.values as StepElementConfig}
                            template={formikProps.values as StepElementConfig}
                            readonly={true}
                            type={(formikProps.values as StepElementConfig)?.type as StepType}
                            stepViewType={StepViewType.InputSet}
                            allowableTypes={allowableTypes}
                          />
                          {getMultiTypeFromValue((formikProps.values as StepElementConfig).spec?.delegateSelectors) ===
                            MultiTypeInputType.RUNTIME && (
                            <div className={cx(stepCss.formGroup, stepCss.sm)}>
                              <MultiTypeDelegateSelector
                                inputProps={{
                                  projectIdentifier: template.projectIdentifier,
                                  orgIdentifier: template.orgIdentifier
                                }}
                                allowableTypes={allowableTypes}
                                label={getString('delegate.DelegateSelector')}
                                name={'spec.delegateSelectors'}
                                disabled={true}
                              />
                            </div>
                          )}
                        </Container>
                      )}
                    </>
                  )
                }}
              </Formik>
            </Layout.Vertical>
          </Container>
        )}
      </Layout.Vertical>
    </Container>
  )
}
