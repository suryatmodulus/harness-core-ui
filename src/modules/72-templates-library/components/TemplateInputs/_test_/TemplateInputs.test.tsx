/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TemplateInputs } from '@templates-library/components/TemplateInputs/TemplateInputs'
import { mockTemplates, mockTemplatesInputYaml } from '@templates-library/TemplatesTestHelper'
import { TestWrapper } from '@common/utils/testUtils'

jest.mock('@common/components/YAMLBuilder/YamlBuilder')
jest.mock('@wings-software/monaco-yaml/lib/esm/languageservice/yamlLanguageService', () => ({
  getLanguageService: jest.fn()
}))

jest.mock('services/template-ng', () => ({
  ...(jest.requireActual('services/template-ng') as any),
  useGetTemplateInputSetYaml: jest
    .fn()
    .mockImplementation(() => ({ data: mockTemplatesInputYaml, refetch: jest.fn(), error: null, loading: false }))
}))

describe('<TemplateInputs /> tests', () => {
  test('snapshot test', async () => {
    const { container } = render(
      <TestWrapper>
        <TemplateInputs template={mockTemplates?.data?.content?.[0] || {}} />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
