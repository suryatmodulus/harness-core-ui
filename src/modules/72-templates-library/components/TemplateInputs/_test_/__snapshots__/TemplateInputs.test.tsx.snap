// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateInputs /> tests snapshot test 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main container StyledProps--padding-top-xlarge StyledProps--padding-left-xxlarge StyledProps--padding-right-xxlarge"
    style="overflow: auto;"
  >
    <div
      class="Layout--vertical StyledProps--main"
    >
      <div
        class="StyledProps--font StyledProps--main inputsContainer"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-xlarge StyledProps--main"
        >
          <div
            class="StyledProps--font StyledProps--main"
          >
            <div
              class="Layout--horizontal Layout--layout-spacing-xxxlarge StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-normal StyledProps--font-weight-bold StyledProps--color StyledProps--color-grey800"
              >
                manju-test-template-qq-12344 (v4)
              </p>
              <p
                class="StyledProps--font StyledProps--main inputsCount StyledProps--font-size-small"
              >
                templatesLibrary.inputsCount
              </p>
            </div>
          </div>
          <div
            class="OverlaySpinner--overlaySpinner"
          >
            <div
              class="StyledProps--font StyledProps--main inputsCard StyledProps--background-white StyledProps--padding-large StyledProps--margin-bottom-xxlarge"
            >
              <div
                class="bp3-form-group bp3-disabled deploymentViewMedium"
              >
                <label
                  class="bp3-label"
                  for="timeout"
                >
                  <span
                    class="Tooltip--acenter"
                    data-tooltip-id="templateInputs_timeout"
                  >
                    pipelineSteps.timeoutLabel
                  </span>
                   
                  <span
                    class="bp3-text-muted"
                  />
                </label>
                <div
                  class="bp3-form-content"
                >
                  <div
                    class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                    style="flex-grow: 1;"
                  >
                    <div
                      class="TextInput--main MultiTypeInput--input"
                    >
                      <div
                        class="bp3-input-group bp3-disabled"
                      >
                        <input
                          class="bp3-input"
                          disabled=""
                          name="timeout"
                          placeholder="<+input>"
                          type="text"
                          value="<+input>"
                        />
                      </div>
                    </div>
                    <button
                      class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                      disabled=""
                    >
                      <span
                        class="bp3-icon StyledProps--main"
                        data-icon="runtime-input"
                      >
                        <svg
                          height="12"
                          viewBox="0 0 10 7.97"
                          width="12"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                            fill="current-color"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="bp3-form-group bp3-disabled approvalMessage deploymentViewMedium"
              >
                <label
                  class="bp3-label"
                  for="spec.approvalMessage"
                >
                  <span
                    class="Tooltip--acenter"
                    data-tooltip-id="templateInputs_spec.approvalMessage"
                  >
                    pipeline.approvalStep.message
                  </span>
                   
                  <span
                    class="bp3-text-muted"
                  />
                </label>
                <div
                  class="bp3-form-content"
                >
                  <div
                    style="display: flex; align-items: center;"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                      style="flex-grow: 1;"
                    >
                      <div
                        class="TextInput--main MultiTypeInput--input"
                      >
                        <div
                          class="bp3-input-group bp3-disabled"
                        >
                          <input
                            class="bp3-input"
                            disabled=""
                            name="spec.approvalMessage"
                            placeholder="<+input>"
                            type="text"
                            value="<+input>"
                          />
                        </div>
                      </div>
                      <button
                        class="MultiTypeInput--btn MultiTypeInput--RUNTIME multiBtn"
                        disabled=""
                      >
                        <span
                          class="bp3-icon StyledProps--main"
                          data-icon="runtime-input"
                        >
                          <svg
                            height="12"
                            viewBox="0 0 10 7.97"
                            width="12"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                              fill="current-color"
                            />
                          </svg>
                        </span>
                      </button>
                    </div>
                    <span
                      class="bp3-popover-wrapper bp3-dark"
                    >
                      <span
                        class="bp3-popover-target"
                      >
                        <button
                          class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withRightIcon"
                          id="configureOptions_spec.approvalMessage"
                          style="margin-bottom: 12px;"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="bp3-icon bp3-icon-cog StyledProps--main"
                            icon="cog"
                          >
                            <svg
                              data-icon="cog"
                              height="16"
                              viewBox="0 0 16 16"
                              width="16"
                            >
                              <desc>
                                cog
                              </desc>
                              <path
                                d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                                fill-rule="evenodd"
                              />
                            </svg>
                          </span>
                        </button>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="deploymentViewMedium"
              >
                <div
                  class="bp3-form-group"
                >
                  <label
                    class="bp3-label"
                  >
                    <span
                      class="Tooltip--acenter"
                      data-tooltip-id="harnessApprovalRuntime_userGroups"
                    >
                      common.userGroups
                    </span>
                     
                    <span
                      class="bp3-text-muted"
                    />
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main multiTypeSupport"
                    >
                      <div
                        class="TextInput--main MultiTypeInput--input"
                      >
                        <div
                          class="bp3-input-group bp3-disabled"
                        >
                          <input
                            class="bp3-input"
                            disabled=""
                            name="spec.approvers.userGroups"
                            placeholder="<+input>"
                            type="text"
                            value="<+input>"
                          />
                        </div>
                      </div>
                      <button
                        class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                        disabled=""
                      >
                        <span
                          class="bp3-icon StyledProps--main"
                          data-icon="runtime-input"
                        >
                          <svg
                            height="12"
                            viewBox="0 0 10 7.97"
                            width="12"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                              fill="current-color"
                            />
                          </svg>
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="bp3-form-group deploymentViewMedium"
              >
                <label
                  class="bp3-label"
                  for="spec.approvers.minimumCount"
                >
                  <span
                    class="Tooltip--acenter"
                    data-tooltip-id="templateInputs_spec.approvers.minimumCount"
                  >
                    pipeline.approvalStep.minimumCount
                  </span>
                   
                  <span
                    class="bp3-text-muted"
                  />
                </label>
                <div
                  class="bp3-form-content"
                >
                  <div
                    class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                  >
                    <div
                      class="TextInput--main MultiTypeInput--input"
                    >
                      <div
                        class="bp3-input-group bp3-disabled"
                      >
                        <input
                          class="bp3-input"
                          disabled=""
                          name="spec.approvers.minimumCount"
                          placeholder="<+input>"
                          type="text"
                          value="<+input>"
                        />
                      </div>
                    </div>
                    <span
                      class="bp3-popover-wrapper MultiTypeInput--wrapper"
                    >
                      <span
                        class="bp3-popover-target"
                      >
                        <button
                          class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                        >
                          <span
                            class="bp3-icon StyledProps--main"
                            data-icon="runtime-input"
                          >
                            <svg
                              height="12"
                              viewBox="0 0 10 7.97"
                              width="12"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                                fill="current-color"
                              />
                            </svg>
                          </span>
                        </button>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
